### ISSUE TITLE
[#8026] Add loading state for currency conversion

### ISSUE THREAD

**Issue by rachelhamlin** [status-im/status-react#8443]
### Description

*Type*: Feature

*Summary*: Continuation of the work done in [this PR](https://github.com/status-im/status-react/pull/8361#issuecomment-500465251) for #8026. 

New overview screen shows fiat values for the total amount + gas. This can sometimes take time to load, so we need to add a loading state.

#### Expected behavior

Designs found here: [preferred implementation](https://github.com/status-im/status-react/issues/8443#issuecomment-504438031), [back-up option](https://github.com/status-im/status-react/issues/8443#issuecomment-528930247) 

When converted currency in the following fields is not yet available, we display a loading state instead.

![image](https://user-images.githubusercontent.com/15470435/59840107-a7493800-9349-11e9-8b43-c4483526eea1.png)
_Loaded currencies_

![image](https://user-images.githubusercontent.com/15470435/64528970-abaa3f00-d301-11e9-88e6-0aec39ca4c69.png)
_Preferred design is an in-line animation (see [commen](https://github.com/status-im/status-react/issues/8443#issuecomment-504438031)t)_

**Comment by hesterbruikman** [comment]
A related issue:

In case of errors, the expected pattern is the "Tooltip error" that has a little tip at the bottom and hovers slightly above the list item that contains the error. In the merged build, the tool misses the tip. It has appeared, in the correct position, in an earlier test build, which is where the image on the right comes from.

cc @andmironov 

![Frame 2](https://user-images.githubusercontent.com/35233054/59859679-7b47aa00-937d-11e9-8144-878c4a4e60f0.png)

**Comment by andmironov** [comment]
I'd suggest something like this to indicate no ETH for gas:
- Using "network fee" rather than "Gas" for consistency;
- Letting know that you have to have the required amount on the wallet you're sending the transaction from;
- Providing a helpful call-to-action.

As for the wrong password, I'd use the default tooltip

<img width="375" alt="no-gas" src="https://user-images.githubusercontent.com/3115606/59924402-f10f4c80-9435-11e9-894c-e94423da334c.png">
<img width="375" alt="wp" src="https://user-images.githubusercontent.com/3115606/59924700-9d513300-9436-11e9-8f5b-2db3d48e550d.png">

**Comment by errorists** [comment]
for the loader I'd use an inline spinner

![inlineLoader](https://user-images.githubusercontent.com/4932774/59927925-de991100-943d-11e9-9eae-5190e6460444.gif)

engineers `activityIndicator, color: grey, size: iOS: small  Android: 16`

**Comment by andmironov** [comment]
FIgured it might be better to re-use this thing to dosplay the warning:
<img width="375" alt="no-eth" src="https://user-images.githubusercontent.com/3115606/60013175-77ba6880-9686-11e9-8a2e-7d4564bd0e57.png">

**Comment by hesterbruikman** [comment]
> 
> 
> FIgured it might be better to re-use this thing to dosplay the warning:
> <img alt="no-eth" width="375" src="https://user-images.githubusercontent.com/3115606/60013175-77ba6880-9686-11e9-8a2e-7d4564bd0e57.png">

Looks like this should be a new issue to update the 'Not enough ETH'-warning. Love the entry to Receive ETH on there!

**Comment by hesterbruikman** [comment]
> 
> 
> FIgured it might be better to re-use this thing to dosplay the warning:
> <img alt="no-eth" width="375" src="https://user-images.githubusercontent.com/3115606/60013175-77ba6880-9686-11e9-8a2e-7d4564bd0e57.png">

Seperate the loading state update (this issue) from the warning update: https://github.com/status-im/status-react/issues/8485

**Comment by hesterbruikman** [comment]
I believe we also need a loading state for the fiat value shown in the wallet overview and individual accounts. It currently shows '0' until conversion is fetched.

@flexsurfer @rachelhamlin should that be a seprate issue or can it be takn along with this one?
cc @andmironov

**Comment by flexsurfer** [comment]
@hesterbruikman can be done here

**Comment by hesterbruikman** [comment]
@andmironov can you share a design here for loading state of fiat value in Account overview and on Account card?

**Comment by hesterbruikman** [comment]
> 
> 
> @andmironov can you share a design here for loading state of fiat value in Account overview and on Account card?

And also what should happen on refresh. Referring to "Refresh wallet should work like it worked before"

**Comment by andmironov** [comment]
>Refresh wallet should work like it worked before

Pull-to-refresh, a spinner is shown while loading:
<img width="392" alt="Screen Shot 2019-07-26 at 13 40 36" src="https://user-images.githubusercontent.com/3115606/61946356-02a2c180-afab-11e9-8f4f-51dc2b131ab6.png">
Here's an old GIF with the old design, but the concept is the same https://dribbble.com/shots/3777816-Status-Wallet

>can you share a design here for loading state of fiat value

Here's what I've had in Figma prepared, made under the impression that it might take a lot of time to load the list of assets and balance:
<img width="361" alt="Screen Shot 2019-07-26 at 13 42 46" src="https://user-images.githubusercontent.com/3115606/61946440-45649980-afab-11e9-9bdc-6f69c5cfe7df.png">

If it's not the case, maybe it would be better to go with something like we did in the Dapp Store (shining placeholders):
<img width="432" alt="Screen Shot 2019-07-26 at 13 43 43" src="https://user-images.githubusercontent.com/3115606/61946497-66c58580-afab-11e9-9526-95b8a64f16b6.png">

up to @flexsurfer to decide which approach fits better technically. Also, the little spinner is used in the transaction overview, which might be another reason not to use the "placeholder" approach

**Comment by hesterbruikman** [comment]
🙌 thanks @andmironov

**Comment by rachelhamlin** [comment]
I'm about to put a bounty on this and want to clarify-

I believe the error tool tip is already fixed. We're also planning on using a different 'not enough ETH' warning in issue #8485, so I'm going to reduce the scope of this issue to just the inline spinner.

Okay for you @andmironov?

**Comment by andmironov** [comment]
Yes, thanks @rachelhamlin ! The scope of this issue (for the transaction overview modal) I believe boils down to the awesome animation @errorists have made! As a fallback solution I can suggest doing:

![Screen Shot 2019-09-06 at 19 51 08](https://user-images.githubusercontent.com/3115606/64445674-b2596c00-d0df-11e9-8d2a-00e38bd2bb0d.png)

I like the animation more, just being careful if it takes more time/is harder to implement

And the error states will look like this: 

![Screen Shot 2019-09-06 at 19 55 23](https://user-images.githubusercontent.com/3115606/64446024-770b6d00-d0e0-11e9-9457-c8f0395fcd74.png)


Note: As far as I understand it's part of https://github.com/status-im/status-react/issues/8485 but it exceeds that issue's scope a bit as well 🤷‍♂️

**Comment by rachelhamlin** [comment]
Okay, awesome, I’ll update and bounty :) thanks for your input.

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 0.472 ETH (99.4 USD @ $210.6/ETH)  attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/status-im/status-react/8443/3471).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/status-im/status-react/8443/3471).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $153,770.40 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Cancelled** 

<hr>

__Work has been started__.


These users each claimed they can complete the work by 5 months ago.
Please review their action plans below:


**1) [acolytec3](https://gitcoin.co/acolytec3) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/status-im/status-react/8443/3471?mutate_worker_action=approve&worker=acolytec3) | [reject worker](https://gitcoin.co/issue/status-im/status-react/8443/3471?mutate_worker_action=reject&worker=acolytec3))_.**

I would like to work on this issue following the one of the proposed implementations linked in the bounty issue.
**2) [rajatkapoor](https://gitcoin.co/rajatkapoor) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/status-im/status-react/8443/3471?mutate_worker_action=approve&worker=rajatkapoor) | [reject worker](https://gitcoin.co/issue/status-im/status-react/8443/3471?mutate_worker_action=reject&worker=rajatkapoor))_.**

I will implement it to exactly match the desired design showing an inline loader
**3) [alexjg](https://gitcoin.co/alexjg) has been approved to start work.**

I've enjoyed working with this codebase, but so far I've just been working on styling things. This looks like a chance to get a bit more contact with the reagent subscriptions framework. I have a bit of time this weekend so would love to have a shot at it.
**4) [morevolk-latei](https://gitcoin.co/morevolk-latei) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/status-im/status-react/8443/3471?mutate_worker_action=approve&worker=morevolk-latei) | [reject worker](https://gitcoin.co/issue/status-im/status-react/8443/3471?mutate_worker_action=reject&worker=morevolk-latei))_.**

I'm good at creating persistent UI's and has a keen eye on frontend design.
I would like to work on this feature and I believe I'm befitting for this.
I have experience in working with React, AngularJs, JS, Node, Django, Linux, etc.
I would love to contribute. :-)

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/status-im/status-react/8443/3471).

**Comment by rachelhamlin** [comment]
Heyo! Sorry for delay on this issue, just got back from our team offsite. 

@alexjg I've approved you :) Welcome back!  Approved based on experience w/ codebase, quality past contribution and first come, first serve.

@acolytec3 will hopefully be contributing on #8027 instead. 

@bshevchenko are you having any better luck getting Status set up after the issue found in #8657? 🤞we can get that fixed.

@rajatkapoor & @morevolk-latei thank you for applying! Do you both have ClojureScript experience? If so, I'm happy to bounty up other UI issues for you to try if interested :)

**Comment by bshevchenko** [comment]
@rachelhamlin Pedro answered me back just a few minutes ago on that issue. Trying now what he advised me to do

**Comment by bshevchenko** [comment]
@rachelhamlin and again no luck :(

**Comment by alexjg** [comment]
@rachelhamlin brilliant, I don't have much bandwidth (literal and mental) this week but I should be able to get this sorted the early part of next week.

**Comment by acolytec3** [comment]
@alexjg FYI, I'd started poking around the code base trying to figure out how this one would work and it requires the one I'm going to try and finish (#8027) before you'll have anything to work with since the currency conversion isn't currently in the code.  Whenever we get there, this is the code that will render the activity indicator they want.  I got that much working already. :-)

```
[react/activity-indicator {
                :color   :colors/gray
                :ios   {:size  :small}
                :android {:size :16}
             }]
```

**Comment by gitcoinbot** [comment]
@alexjg Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=1) | [3 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=3) | [5 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=5) | [10 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=10) | [100 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=100)

**Comment by gitcoinbot** [comment]
@alexjg Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=1) | [3 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=3) | [5 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=5) | [10 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=10) | [100 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=100)

**Comment by alexjg** [comment]
@gitcoinbot I'm still planning on working on this, although it looks like it may have already been resolved in #8027

**Comment by hesterbruikman** [comment]
@alexjg Looks like the in line spinner is implemented in #9063, I'm missing the total value loader @andmironov can you check what remains to be implemented of the loading state design that @alexjg can work on here?

**Comment by andmironov** [comment]
> @andmironov can you check what remains to be implemented of the loading state design 

cc @rachelhamlin i'm not sure where I should be looking for the latest build to check that. 

The only other loading state of the transaction overview I can remember is "Calculating..." the network fee when interaction with a contract. Do you happen to know if that has been addressed already? I believe it's part of the "error states" issue now, isn't it?

**Comment by hesterbruikman** [comment]
Seems like the loading state for the Total value and the 'pull to refresh' loading state is not implemented yet. 

To test:
Install build from #9063 (https://github.com/status-im/status-react/pull/9063#issuecomment-536079760)
- Open wallet 
- Check Total value
- Switch to airplane mode
- Tap on `...` and `set currency`
- Select another currency
- Check total value (will now be indicated as `0`

@andmironov I imagine for the total value we would include [this version](https://user-images.githubusercontent.com/3115606/61946497-66c58580-afab-11e9-9526-95b8a64f16b6.png) you commented above. 

However, having this state show indefinitely doesn't seem great. We might need a variation to represent the Offline state. Either the good old snackbar, that I don't see implemented here, or another indicator that says values are not 'live'. Wdyt?

@alexjg I would create a seperate issue for this offline state. For now I'd say this issue includes:
- [ ] Pull to refresh
- [ ] Total value loading state
Does that work for you?

cc @rachelhamlin

**Comment by andmironov** [comment]
>I imagine for the total value we would include this version you commented above.

sorry i'd go with this if it's possible

<img width="361" alt="Screen Shot 2019-07-26 at 13 42 46" src="https://user-images.githubusercontent.com/3115606/61946440-45649980-afab-11e9-9bdc-6f69c5cfe7df.png">

**Comment by rachelhamlin** [comment]
Hey @alexjg - apologies that this issue was out of date. Are you still tracking? I'm happy to refocus this on the items from @hesterbruikman's [comment](https://github.com/status-im/status-react/issues/8443#issuecomment-537469823).

@hesterbruikman pull to refresh = on wallet overview, and individual wallet?

**Comment by hesterbruikman** [comment]
Good point. Ideally loading state for wallet overview and individual account screen @andmironov can you point to the design of the loading state for the individual account?

Regarding pull to refresh, @andmironov, wdyt? Not sure how this would behave with the title/header on the individual account screen.

**Comment by alexjg** [comment]
@rachelhamlin I've been keeping half an eye on this, still keen to work on it. @andmironov suggested creating a new issue, do you want to do that or continue with this one?

**Comment by rachelhamlin** [comment]
It sounds like the original issue is no longer relevant, but since there's already a bounty attached, I can repurpose it according to the comments :) I'll update the issue once we have alignment. OK for you @alexjg? 

1. Pull to refresh behavior is uncertain - when and how should this work? (@andmironov task)
2. Loading state should be implemented for multiaccount overview, per [this comment](https://github.com/status-im/status-react/issues/8443#issuecomment-537975995) (Figma?)
3. Loading state for individual accounts should be implemented as well - Figma? 

@hesterbruikman cc'ing you to sanity check me!

**Comment by alexjg** [comment]
Works for me, I'll wait for you folks to clarify things before I get started.

**Comment by andmironov** [comment]
Sorry for a slight change of plans, but after diving a bit deeper into this issue i realized that it's better not to use animated placeholders here, since this screen is not exactly a use-case for those. For instance, it would be pretty annoying if you want just to update the conversion rate but instead you get all your list of assets replaced with placeholders.

And the alternative is a good old spinner which also should be easier to implement. I am not quite sure if there's a "loading spinner" component somewhere in our codebase available, but it it's not the case, it'd better to be implemented as a re-seable element here. cc @flexsurfer Or if it takes too much time we cal always use a native spinner (temporary) 

**Multiaccount overview** 

It has two pieces: pull-to-refresh spinner and the loading state. Also, there should be an error state. It's better to create a prototype to illustrate how it should work exactly, but the rough idea is: 

1. The first time the user opens the wallet they see the loading animation

<img width="420" alt="Screen Shot 2019-10-07 at 16 28 37" src="https://user-images.githubusercontent.com/3115606/66320695-9e439d00-e91f-11e9-9af1-b0d8497b4d9e.png">

2. After the loading has finished, the spinner disappears, the screen is "scrolled" back up. Here's a prototype I've made ages ago for the old wallet design https://dribbble.com/shots/3777816-Status-Wallet Notice how once the loading is finished, the spinner disappears and the screen moves up.

3. If an error happens during loading, there's an error message with a "Retry" button. 

<img width="408" alt="Screen Shot 2019-10-07 at 16 31 26" src="https://user-images.githubusercontent.com/3115606/66320868-ec58a080-e91f-11e9-9e38-02811df28bbb.png">

The animation of how the spinner disappears and is replaced by the error message, and how a new spinner is shown when user taps "retry" should be simple, I'd suggest translating size and opacity from 1 to 0 and vice versa. Notice that the screen is not moving up or down when the spinner is replaced with the error message.

<img width="1506" alt="Screen Shot 2019-10-07 at 16 33 58" src="https://user-images.githubusercontent.com/3115606/66321073-49545680-e920-11e9-9c10-ccd497adff52.png">


4. If the user pulls up, a spinner appears. It must appear gradually, using modulation. Just like on the prototype above: the more the screen is pulled down, the closer the spinner is to its final state. If the user stops before a "final" point, the refresh does not happen. I'd also suggest translating size and opacity from 1 to 0.

<img width="560" alt="Screen Shot 2019-10-07 at 16 36 24" src="https://user-images.githubusercontent.com/3115606/66321286-a2bc8580-e920-11e9-850b-ad571209a305.png">

All the Figma files are right here https://www.figma.com/file/XUehMnhyD1FGcWzvGz6SXqvh/Mobile-wallet?node-id=1743%3A303

Please notice that we use a design system (we re-use colors, icons, text styles, component) which is located here https://www.figma.com/file/cb4p8AxLtTF3q1L6JYDnKN15/Index?node-id=22%3A0

**Comment by andmironov** [comment]
**Individual accounts**

Initial loading and pull-to-refresh works almost exactly the same as the account explorer

<img width="841" alt="Screen Shot 2019-10-07 at 17 05 37" src="https://user-images.githubusercontent.com/3115606/66323842-b669eb00-e924-11e9-8d0f-98fdf30b38cb.png">


But the thing is that when the history tab is opened for the first time, it has a loading state of its own since it would not be wise to pre-load it, as it's the third tab in the UI and it's not the primary information on the individual account screen. The history list might be really long and it might take a long time to load that list. This is where the animated placeholders might be useful, but for simplicity's sake we can do something like this:

<img width="836" alt="Screen Shot 2019-10-07 at 17 04 49" src="https://user-images.githubusercontent.com/3115606/66323767-96d2c280-e924-11e9-9fc2-0dcb5b9008f8.png">


The history list has another loading state when it's scrolled to the bottom (scroll-initiated pagination) I'd suggest initially loading only the last 10 items (might change depending on UX testing)

<img width="411" alt="Screen Shot 2019-10-07 at 17 04 13" src="https://user-images.githubusercontent.com/3115606/66323722-80c50200-e924-11e9-9754-3497afa036f5.png">

The Figma files are located here https://www.figma.com/file/XUehMnhyD1FGcWzvGz6SXqvh/Mobile-wallet?node-id=1743%3A303

**Comment by gitcoinbot** [comment]
@alexjg Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=1) | [3 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=3) | [5 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=5) | [10 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=10) | [100 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=100)

**Comment by gitcoinbot** [comment]
@alexjg Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=1) | [3 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=3) | [5 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=5) | [10 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=10) | [100 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=100)

**Comment by alexjg** [comment]
@gitcoinbot I'm still working on this. I've been waiting for the spec to shake out. which looks like it's happened now so I'll look to make a start on it tomorrow.

**Comment by gitcoinbot** [comment]
@alexjg Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=1) | [3 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=3) | [5 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=5) | [10 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=10) | [100 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=100)

**Comment by gitcoinbot** [comment]
@alexjg Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=1) | [3 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=3) | [5 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=5) | [10 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=10) | [100 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=100)

**Comment by alexjg** [comment]
@gitcoinbot I'm still working on it, going slow due to life intruding on my quiet moments of software engineering. Looking to get it finished this week though.

**Comment by gitcoinbot** [comment]
@alexjg Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=1) | [3 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=3) | [5 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=5) | [10 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=10) | [100 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=100)

**Comment by gitcoinbot** [comment]
@alexjg Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=1) | [3 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=3) | [5 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=5) | [10 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=10) | [100 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=100)

**Comment by StatusSceptre** [comment]
@alexjg pretty sure you're still making progress in your own time but check in w/ us this week if you can :)

**Comment by alexjg** [comment]
Apologies, I thought I had responded to @gitcoinbot. Yup I'm still trucking on, hoping to submit a PR tomorrow.

**Comment by gitcoinbot** [comment]
@alexjg Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=1) | [3 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=3) | [5 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=5) | [10 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=10) | [100 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=100)

**Comment by gitcoinbot** [comment]
@alexjg Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=1) | [3 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=3) | [5 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=5) | [10 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=10) | [100 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=100)

**Comment by StatusSceptre** [comment]
Hey @alexjg - checking in, I bet you're still working but want to confirm.

**Comment by alexjg** [comment]
@StatusSceptre @gitcoinbot I'm still working. Life has been very hectic the last few weeks, hence low productivity. Apologies for that. I am aiming to finish this one off ASAP.

**Comment by StatusSceptre** [comment]
All good @alexjg! Good to know. Hope all's well.

**Comment by gitcoinbot** [comment]
@alexjg Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=1) | [3 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=3) | [5 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=5) | [10 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=10) | [100 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=100)

**Comment by gitcoinbot** [comment]
@alexjg Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=1) | [3 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=3) | [5 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=5) | [10 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=10) | [100 days](https://gitcoin.co/issue/status-im/status-react/8443/3471?snooze=100)

**Comment by alexjg** [comment]
@gitcoinbot Yup still going. Pushing hard to get a PR in before the weekend.

**Comment by alexjg** [comment]
Quick update here. I'm working on getting the `refreshControl` property of the relevant `ScrollView`s set up in order to build the specced UI. I'm struggling a little to figure out how to pass a react component as a property to a reagent view, which I need to do to pass `RefreshControl` to the `refreshControl` property, but I'll figure it out.

In terms of events and applications state I'm thinking to display the loading spinner whenever `:prices-loading?` is `true` and to dispatch `:wallet-ui.pull-to-refresh` when pulling to refresh, hopefully that makes sense.

As a side note I'm running into some issues with the build environment, specifically I have to completely restart the application if I run into any errors. Additionally, tapping on a yellow error box causes the application to crash, so not only do I have to restart the application but I can't easily get access to compile errors. I'll chase this up in other avenues though.

**Comment by StatusSceptre** [comment]
Hey @alexjg - do you need any help trouble shooting your build env?

**Comment by alexjg** [comment]
Hey @StatusSceptre possibly yeah. Should I create issues on the bug tracker or ask for advice in discuss.status.im?

**Comment by alexjg** [comment]
I'm working around it at the moment. It's not slowing me down too much as I'm mostly stuck on how to get the `refreshControl` property to actually do anything. I've successfully passed the `RefreshControl` to the `ScrollView` using `reagent/as-element` but it's not doing anything useful (`ScrollView` just displays blank). I'm still troubleshooting that but am beginning to wonder if I'll need to write the pull down to refresh functionality from scratch.

**Comment by alexjg** [comment]
I've managed to get `RefreshControl` working but it doesn't really give us the hooks we need. It renders the spinner on top of the scrollview content in the z axis and doesn't really expose any way to render the spinner elsewhere or slowly transition it in as the user scrolls down. I think we'll need to build the pull to refresh functionality ourselves, so unless there's already something in the codebase to do that (I can't find anything) then I'll make a start on it tomorrow.

**Comment by StatusSceptre** [comment]
I'm just going to ping a few people who might be able to help. @alexjg is having trouble with his build environment - @PombeirP @yenda @flexsurfer could one of you possibly help trouble shoot?

**Comment by flexsurfer** [comment]
@alexjg @rachelhamlin i'm not sure , issue title says "Add loading state for currency conversion" why do we need RefreshControl for it?

**Comment by flexsurfer** [comment]
> As a side note I'm running into some issues with the build environment, specifically I have to completely restart the application if I run into any errors. Additionally, tapping on a yellow error box causes the application to crash, so not only do I have to restart the application but I can't easily get access to compile errors. I'll chase this up in other avenues though.

@alexjg could you provide more details pls better in a new issue with screenshots or video thanks!

**Comment by alexjg** [comment]
@flexsurfer I've created a new issue [here](https://github.com/status-im/status-react/issues/9432)

**Comment by alexjg** [comment]
@flexsurfer as regards the RefreshControl, there's been a slight change in scope throughout this issue and the plan is now to implement a pull to refresh interaction for the currency conversion rather than an inline loading state. See [this comment](https://github.com/status-im/status-react/issues/8443#issuecomment-539045418)

**Comment by flexsurfer** [comment]
hey @alexjg i've discussed pull and refresh with @andmironov we can ignore it for this issue, and just show loader indicator when we load balance

**Comment by alexjg** [comment]
As in just show the loading spinner at the top of the page (and associated error report) but not implement the pull to refresh?

**Comment by flexsurfer** [comment]
@alexjg correct, thanks!

**Comment by andmironov** [comment]
Hey! Just to make things clear, I still think having a pull-to-refresh is important, but it could be ignored **for now** to keep things simple

> just show the loading spinner at the top of the page (and associated error report) but not implement the pull to refresh

that would work for now! I don't know how often the amounts get updated in the background, but if we want to keep things transparent and stay resourceful (Principles!) we still need to show this after a certain timeout:

<img width="405" alt="Screen Shot 2019-11-14 at 12 19 03" src="https://user-images.githubusercontent.com/3115606/68843326-f74de000-06d8-11ea-93d0-df400d64d857.png">

**Comment by alexjg** [comment]
I'm happy to add that UI to the component I'm building, it's not particularly complicated. However, where's the best place to track that timeout? As part of the component state or is there something in the application state I can use?

**Comment by andmironov** [comment]
sorry, I am not familiar with how often the amounts get updated and how long it takes so let's find that out first :-) maybe there's no reason for this state to exist at all

**Comment by hesterbruikman** [comment]
>

**Comment by hesterbruikman** [comment]
> how often amounts get updated and how long it takes.

This is a good point. We used to have issues with cryptocompare and a far from pretty loading state would persists, covering the wallet amount. But maybe the conversion pulling works different now.

**Comment by yenda** [comment]
@alexjg you can call `update-prices` https://github.com/status-im/status-react/blob/88dbeead42f640a1ba8592041bbdcff656297799/src/status_im/wallet/core.cljs#L242

It uses `get-prices` here https://github.com/status-im/status-react/blob/298f000fcb4bd98b4180f196fcd80c0508696983/src/status_im/utils/prices.cljs#L42 to which you can add a timeout

**Comment by alexjg** [comment]
I've been using the `prices-loading` and `prices-update` state to track most of this stuff at the moment so adding a timeout shouldn't be too tricky. Most of what I'm doing at the moment is figuring out how the react native animations framework works in a reagent context. I've pretty much got my head around that now so I'm just putting it all together.

**Comment by StatusSceptre** [comment]
Hey @alexjg - just a regular check-in. I'm surprised Gitcoinbot hasn't been pinging you! How are things going?

**Comment by alexjg** [comment]
Hey! Apologies, this dropped off my radar with the encroachment of christmas. I should have time to complete it over the holidays though.

**Comment by rachelhamlin** [comment]
Cool, no sweat @alexjg. We'll consider you still active :)

**Comment by alexjg** [comment]
@rachelhamlin I think realistically I'm not going to have the time to complete this work. I've submitted a PR with the work I've done so far. The loading spinner UI (the layout of the loading notification and animation of the spinner) is done, but I haven't had a chance to complete the logic to animate showing and hiding the loading spinner as the application state changes. It shouldn't be too tough to do, unfortunately I just have too much on my plate to get to it. Apologies for dragging this out, I really wanted to contribute but don't have the bandwitdth right now.

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Cancelled** 

<hr>

__The funding of 0.472 ETH  (55.27 USD @ $117.09/ETH) attached to this issue has been **cancelled** by the bounty submitter__

 
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $80,464.65 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by rachelhamlin** [comment]
@vkjr awesome work!
